// arktype/dist/cjs/utils/data.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    sizeOf: ()=>sizeOf,
    unitsOf: ()=>unitsOf,
    DataWrapper: ()=>DataWrapper
});
const _domainsJs = require("./domains.js");
const _serializeJs = require("./serialize.js");
function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
const sizeOf = (data)=>typeof data === "string" || Array.isArray(data) ? data.length : typeof data === "number" ? data : 0;
const unitsOf = (data)=>typeof data === "string" ? "characters" : Array.isArray(data) ? "items long" : "";
class DataWrapper {
    toString() {
        return (0, _serializeJs.stringify)(this.value);
    }
    get domain() {
        return (0, _domainsJs.domainOf)(this.value);
    }
    get size() {
        return sizeOf(this.value);
    }
    get units() {
        return unitsOf(this.value);
    }
    get className() {
        return Object(this.value).constructor.name;
    }
    constructor(value){
        _defineProperty(this, "value", void 0);
        this.value = value;
    }
}
