// arktype/dist/cjs/utils/serialize.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    snapshot: ()=>snapshot,
    stringify: ()=>stringify,
    serializePrimitive: ()=>serializePrimitive,
    deserializePrimitive: ()=>deserializePrimitive
});
const _domainsJs = require("./domains.js");
const _genericsJs = require("./generics.js");
const _numericLiteralsJs = require("./numericLiterals.js");
const snapshot = (data, opts = {})=>serializeRecurse(data, opts, []);
const stringify = (data, indent)=>{
    switch((0, _domainsJs.domainOf)(data)){
        case "object":
            return JSON.stringify(serializeRecurse(data, stringifyOpts, []), null, indent);
        case "symbol":
            return stringifyOpts.onSymbol(data);
        default:
            return serializePrimitive(data);
    }
};
const stringifyOpts = {
    onCycle: ()=>"(cycle)",
    onSymbol: (v)=>`(symbol${v.description && ` ${v.description}`})`,
    onFunction: (v)=>`(function${v.name && ` ${v.name}`})`
};
const serializeRecurse = (data, context, seen)=>{
    switch((0, _domainsJs.domainOf)(data)){
        case "object":
            if (typeof data === "function") {
                return stringifyOpts.onFunction(data);
            }
            if (seen.includes(data)) {
                return "(cycle)";
            }
            const nextSeen = [
                ...seen,
                data
            ];
            if (Array.isArray(data)) {
                return data.map((item)=>serializeRecurse(item, context, nextSeen));
            }
            const result = {};
            for(const k in data){
                result[k] = serializeRecurse(data[k], context, nextSeen);
            }
            return result;
        case "symbol":
            return stringifyOpts.onSymbol(data);
        case "bigint":
            return `${data}n`;
        case "undefined":
            return "undefined";
        default:
            return data;
    }
};
const serializePrimitive = (value)=>typeof value === "string" ? `'${value}'` : typeof value === "bigint" ? `${value}n` : `${value}`;
const deserializePrimitive = (serialized)=>(0, _genericsJs.isKeyOf)(serialized, serializedKeywords) ? serializedKeywords[serialized] : serialized[0] === "'" ? serialized.slice(1, -1) : (0, _numericLiteralsJs.tryParseWellFormedBigint)(serialized) ?? (0, _numericLiteralsJs.tryParseWellFormedNumber)(serialized, true);
const serializedKeywords = {
    true: true,
    false: false,
    undefined,
    null: null
};
