// @effect/schema/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TreeFormatter = exports.Serializable = exports.Schema = exports.Pretty = exports.ParseResult = exports.JSONSchema = exports.FastCheck = exports.Equivalence = exports.ArrayFormatter = exports.Arbitrary = exports.AST = void 0;
var _AST = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./AST.js"));
exports.AST = _AST;
var _Arbitrary = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Arbitrary.js"));
exports.Arbitrary = _Arbitrary;
var _ArrayFormatter = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ArrayFormatter.js"));
exports.ArrayFormatter = _ArrayFormatter;
var _Equivalence = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Equivalence.js"));
exports.Equivalence = _Equivalence;
var _FastCheck = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./FastCheck.js"));
exports.FastCheck = _FastCheck;
var _JSONSchema = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./JSONSchema.js"));
exports.JSONSchema = _JSONSchema;
var _ParseResult = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ParseResult.js"));
exports.ParseResult = _ParseResult;
var _Pretty = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Pretty.js"));
exports.Pretty = _Pretty;
var _Schema = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Schema.js"));
exports.Schema = _Schema;
var _Serializable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Serializable.js"));
exports.Serializable = _Serializable;
var _TreeFormatter = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TreeFormatter.js"));
exports.TreeFormatter = _TreeFormatter;
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
