// @sinclair/typebox/build/import/value/clone/clone.mjs
// ------------------------------------------------------------------
// ValueGuard
// ------------------------------------------------------------------
import { IsArray, IsDate, IsStandardObject, IsTypedArray, IsValueType } from '../guard/index.mjs';
// ------------------------------------------------------------------
// Clonable
// ------------------------------------------------------------------
function ObjectType(value) {
    const keys = [...Object.getOwnPropertyNames(value), ...Object.getOwnPropertySymbols(value)];
    return keys.reduce((acc, key) => ({ ...acc, [key]: Clone(value[key]) }), {});
}
function ArrayType(value) {
    return value.map((element) => Clone(element));
}
function TypedArrayType(value) {
    return value.slice();
}
function DateType(value) {
    return new Date(value.toISOString());
}
function ValueType(value) {
    return value;
}
// ------------------------------------------------------------------
// Clone
// ------------------------------------------------------------------
/** Returns a clone of the given value */
export function Clone(value) {
    if (IsArray(value))
        return ArrayType(value);
    if (IsDate(value))
        return DateType(value);
    if (IsStandardObject(value))
        return ObjectType(value);
    if (IsTypedArray(value))
        return TypedArrayType(value);
    if (IsValueType(value))
        return ValueType(value);
    throw new Error('ValueClone: Unable to clone value');
}
