// @ark/schema/out/roots/domain.js
import { domainDescriptions, domainOf, getBaseDomainKeys } from "@ark/util";
import { Disjoint } from "../shared/disjoint.js";
import { implementNode } from "../shared/implement.js";
import { InternalBasis } from "./basis.js";
const implementation = implementNode({
    kind: "domain",
    hasAssociatedError: true,
    collapsibleKey: "domain",
    keys: {
        domain: {}
    },
    normalize: schema => typeof schema === "string" ? { domain: schema } : schema,
    defaults: {
        description: node => domainDescriptions[node.domain],
        actual: data => (typeof data === "boolean" ? `${data}` : domainOf(data))
    },
    intersections: {
        domain: (l, r) => Disjoint.init("domain", l, r)
    }
});
export class DomainNode extends InternalBasis {
    traverseAllows = data => domainOf(data) === this.domain;
    compiledCondition = this.domain === "object" ?
        `((typeof data === "object" && data !== null) || typeof data === "function")`
        : `typeof data === "${this.domain}"`;
    compiledNegation = this.domain === "object" ?
        `((typeof data !== "object" || data === null) && typeof data !== "function")`
        : `typeof data !== "${this.domain}"`;
    expression = this.domain;
    literalKeys = getBaseDomainKeys(this.domain);
    get shortDescription() {
        return domainDescriptions[this.domain];
    }
}
export const Domain = {
    implementation,
    Node: DomainNode
};
