// @sinclair/typebox/build/import/value/guard/guard.mjs
// --------------------------------------------------------------------------
// Iterators
// --------------------------------------------------------------------------
/** Returns true if this value is an async iterator */
export function IsAsyncIterator(value) {
    return IsObject(value) && Symbol.asyncIterator in value;
}
/** Returns true if this value is an iterator */
export function IsIterator(value) {
    return IsObject(value) && Symbol.iterator in value;
}
// --------------------------------------------------------------------------
// Nominal
// --------------------------------------------------------------------------
/** Returns true if this value is a typed array */
export function IsTypedArray(value) {
    return ArrayBuffer.isView(value);
}
/** Returns true if this value is a Promise */
export function IsPromise(value) {
    return value instanceof Promise;
}
/** Returns true if the value is a Uint8Array */
export function IsUint8Array(value) {
    return value instanceof Uint8Array;
}
/** Returns true if this value is a Date */
export function IsDate(value) {
    return value instanceof Date && Number.isFinite(value.getTime());
}
// --------------------------------------------------------------------------
// Standard
// --------------------------------------------------------------------------
/** Returns true if this value has this property key */
export function HasPropertyKey(value, key) {
    return key in value;
}
/** Returns true if this object is not an instance of any other type */
export function IsPlainObject(value) {
    return IsObject(value) && IsFunction(value.constructor) && value.constructor.name === 'Object';
}
/** Returns true of this value is an object type */
export function IsObject(value) {
    return value !== null && typeof value === 'object';
}
/** Returns true if this value is an array, but not a typed array */
export function IsArray(value) {
    return Array.isArray(value) && !ArrayBuffer.isView(value);
}
/** Returns true if this value is an undefined */
export function IsUndefined(value) {
    return value === undefined;
}
/** Returns true if this value is an null */
export function IsNull(value) {
    return value === null;
}
/** Returns true if this value is an boolean */
export function IsBoolean(value) {
    return typeof value === 'boolean';
}
/** Returns true if this value is an number */
export function IsNumber(value) {
    return typeof value === 'number';
}
/** Returns true if this value is an integer */
export function IsInteger(value) {
    return IsNumber(value) && Number.isInteger(value);
}
/** Returns true if this value is bigint */
export function IsBigInt(value) {
    return typeof value === 'bigint';
}
/** Returns true if this value is string */
export function IsString(value) {
    return typeof value === 'string';
}
/** Returns true if this value is a function */
export function IsFunction(value) {
    return typeof value === 'function';
}
/** Returns true if this value is a symbol */
export function IsSymbol(value) {
    return typeof value === 'symbol';
}
/** Returns true if this value is a value type such as number, string, boolean */
export function IsValueType(value) {
    // prettier-ignore
    return (IsBigInt(value) ||
        IsBoolean(value) ||
        IsNull(value) ||
        IsNumber(value) ||
        IsString(value) ||
        IsSymbol(value) ||
        IsUndefined(value));
}
