// arktype/out/ark.js
import { arkGenerics } from "./keywords/arkGenerics.js";
import { formatting } from "./keywords/format.js";
import { internal } from "./keywords/internal.js";
import { jsObjects } from "./keywords/jsObjects.js";
import { parsing } from "./keywords/parsing.js";
import { platformObjects } from "./keywords/platformObjects.js";
import { tsGenerics } from "./keywords/tsGenerics.js";
import { tsKeywords } from "./keywords/tsKeywords.js";
import { typedArray } from "./keywords/typedArray.js";
import { validation } from "./keywords/validation.js";
import { scope } from "./scope.js";
export const ambient = scope({
    ...tsKeywords,
    ...jsObjects,
    ...platformObjects,
    ...validation,
    ...internal,
    ...tsGenerics,
    ...arkGenerics,
    TypedArray: typedArray,
    parse: parsing,
    format: formatting
}, { prereducedAliases: true, ambient: true });
export const ark = ambient.export();
export const type = ambient.type;
export const generic = ambient.generic;
export const define = ambient.define;
export const declare = ambient.declare;
