// arktype/dist/cjs/parse/string/reduce/dynamic.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DynamicState", {
    enumerable: true,
    get: ()=>DynamicState
});
const _nodeJs = require("../../../nodes/node.js");
const _rangeJs = require("../../../nodes/rules/range.js");
const _errorsJs = require("../../../utils/errors.js");
const _genericsJs = require("../../../utils/generics.js");
const _serializeJs = require("../../../utils/serialize.js");
const _scannerJs = require("../shift/scanner.js");
const _sharedJs = require("./shared.js");
function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
class DynamicState {
    error(message) {
        return (0, _errorsJs.throwParseError)(message);
    }
    hasRoot() {
        return this.root !== undefined;
    }
    resolveRoot() {
        this.assertHasRoot();
        return this.ctx.type.scope.resolveTypeNode(this.root);
    }
    rootToString() {
        this.assertHasRoot();
        return (0, _serializeJs.stringify)(this.root);
    }
    ejectRootIfLimit() {
        this.assertHasRoot();
        const resolution = typeof this.root === "string" ? this.ctx.type.scope.resolveNode(this.root) : this.root;
        if ((0, _nodeJs.isLiteralNode)(resolution, "number")) {
            const limit = resolution.number.value;
            this.root = undefined;
            return limit;
        }
    }
    ejectRangeIfOpen() {
        if (this.branches.range) {
            const range = this.branches.range;
            delete this.branches.range;
            return range;
        }
    }
    assertHasRoot() {
        if (this.root === undefined) {
            return (0, _errorsJs.throwInternalError)("Unexpected interaction with unset root");
        }
    }
    assertUnsetRoot() {
        if (this.root !== undefined) {
            return (0, _errorsJs.throwInternalError)("Unexpected attempt to overwrite root");
        }
    }
    setRoot(node) {
        this.assertUnsetRoot();
        this.root = node;
    }
    rootToArray() {
        this.root = (0, _nodeJs.toArrayNode)(this.ejectRoot());
    }
    intersect(node) {
        this.root = (0, _nodeJs.rootIntersection)(this.ejectRoot(), node, this.ctx.type);
    }
    ejectRoot() {
        this.assertHasRoot();
        const root = this.root;
        this.root = undefined;
        return root;
    }
    ejectFinalizedRoot() {
        this.assertHasRoot();
        const root = this.root;
        this.root = ejectedProxy;
        return root;
    }
    finalize() {
        if (this.groups.length) {
            return this.error(_sharedJs.unclosedGroupMessage);
        }
        this.finalizeBranches();
        this.scanner.finalized = true;
    }
    reduceLeftBound(limit, comparator) {
        const invertedComparator = _scannerJs.Scanner.invertedComparators[comparator];
        if (!(0, _genericsJs.isKeyOf)(invertedComparator, _rangeJs.minComparators)) {
            return this.error((0, _sharedJs.writeUnpairableComparatorMessage)(comparator));
        }
        if (this.branches.range) {
            return this.error((0, _sharedJs.writeMultipleLeftBoundsMessage)(`${this.branches.range.limit}`, this.branches.range.comparator, `${limit}`, invertedComparator));
        }
        this.branches.range = {
            limit,
            comparator: invertedComparator
        };
    }
    finalizeBranches() {
        this.assertRangeUnset();
        if (this.branches.union) {
            this.pushRootToBranch("|");
            this.setRoot(this.branches.union);
        } else if (this.branches.intersection) {
            this.setRoot((0, _nodeJs.rootIntersection)(this.branches.intersection, this.ejectRoot(), this.ctx.type));
        }
    }
    finalizeGroup() {
        this.finalizeBranches();
        const topBranchState = this.groups.pop();
        if (!topBranchState) {
            return this.error((0, _sharedJs.writeUnmatchedGroupCloseMessage)(this.scanner.unscanned));
        }
        this.branches = topBranchState;
    }
    pushRootToBranch(token) {
        this.assertRangeUnset();
        this.branches.intersection = this.branches.intersection ? (0, _nodeJs.rootIntersection)(this.branches.intersection, this.ejectRoot(), this.ctx.type) : this.ejectRoot();
        if (token === "|") {
            this.branches.union = this.branches.union ? (0, _nodeJs.rootUnion)(this.branches.union, this.branches.intersection, this.ctx.type) : this.branches.intersection;
            delete this.branches.intersection;
        }
    }
    assertRangeUnset() {
        if (this.branches.range) {
            return this.error((0, _sharedJs.writeOpenRangeMessage)(`${this.branches.range.limit}`, this.branches.range.comparator));
        }
    }
    reduceGroupOpen() {
        this.groups.push(this.branches);
        this.branches = {};
    }
    previousOperator() {
        return this.branches.range?.comparator ?? this.branches.intersection ? "&" : this.branches.union ? "|" : undefined;
    }
    shiftedByOne() {
        this.scanner.shift();
        return this;
    }
    constructor(def, ctx){
        _defineProperty(this, "ctx", void 0);
        _defineProperty(this, "scanner", void 0);
        _defineProperty(this, "root", void 0);
        _defineProperty(this, "branches", void 0);
        _defineProperty(this, "groups", void 0);
        this.ctx = ctx;
        this.branches = {};
        this.groups = [];
        this.scanner = new _scannerJs.Scanner(def);
    }
}
const ejectedProxy = new Proxy({}, {
    get: ()=>(0, _errorsJs.throwInternalError)(`Unexpected attempt to access ejected attributes`)
});
