// @effect/schema/dist/cjs/internal/errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRequiredElementFollowinAnOptionalElementErrorMessage = exports.getPrettyErrorMessage = exports.getIndexSignatureParameterErrorMessage = exports.getEquivalenceErrorMessage = exports.getDuplicatePropertySignatureTransformationErrorMessage = exports.getDuplicatePropertySignatureErrorMessage = exports.getDuplicateIndexSignatureErrorMessage = exports.getArbitraryErrorMessage = exports.getAPIErrorMessage = void 0;
var util_ = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./util.js"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
/** @internal */
const getDuplicatePropertySignatureErrorMessage = name => `Duplicate property signature ${util_.formatUnknown(name)}`;
/** @internal */
exports.getDuplicatePropertySignatureErrorMessage = getDuplicatePropertySignatureErrorMessage;
const getDuplicateIndexSignatureErrorMessage = name => `Duplicate index signature for type \`${name}\``;
/** @internal */
exports.getDuplicateIndexSignatureErrorMessage = getDuplicateIndexSignatureErrorMessage;
const getIndexSignatureParameterErrorMessage = exports.getIndexSignatureParameterErrorMessage = "An index signature parameter type must be `string`, `symbol`, a template literal type or a refinement of the previous types";
/** @internal */
const getRequiredElementFollowinAnOptionalElementErrorMessage = exports.getRequiredElementFollowinAnOptionalElementErrorMessage = "A required element cannot follow an optional element. ts(1257)";
/** @internal */
const getDuplicatePropertySignatureTransformationErrorMessage = name => `Duplicate property signature transformation ${util_.formatUnknown(name)}`;
/** @internal */
exports.getDuplicatePropertySignatureTransformationErrorMessage = getDuplicatePropertySignatureTransformationErrorMessage;
const getArbitraryErrorMessage = message => `cannot build an Arbitrary for ${message}`;
/** @internal */
exports.getArbitraryErrorMessage = getArbitraryErrorMessage;
const getPrettyErrorMessage = message => `cannot build a Pretty for ${message}`;
/** @internal */
exports.getPrettyErrorMessage = getPrettyErrorMessage;
const getEquivalenceErrorMessage = message => `cannot build an Equivalence for ${message}`;
/** @internal */
exports.getEquivalenceErrorMessage = getEquivalenceErrorMessage;
const getAPIErrorMessage = (api, message) => `${api}: ${message}`;
exports.getAPIErrorMessage = getAPIErrorMessage;
