// @effect/schema/dist/cjs/internal/filters.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MinLengthTypeId = exports.MinItemsTypeId = exports.MaxLengthTypeId = exports.MaxItemsTypeId = exports.LessThanTypeId = exports.LessThanOrEqualToTypeId = exports.LessThanOrEqualToBigIntTypeId = exports.LessThanBigIntTypeId = exports.LengthTypeId = exports.ItemsCountTypeId = exports.IntTypeId = exports.GreaterThanTypeId = exports.GreaterThanOrEqualToTypeId = exports.GreaterThanOrEqualToBigIntTypeId = exports.GreaterThanBigintTypeId = exports.BetweenTypeId = exports.BetweenBigintTypeId = void 0;
/** @internal */
const GreaterThanTypeId = exports.GreaterThanTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/GreaterThan");
/** @internal */
const GreaterThanOrEqualToTypeId = exports.GreaterThanOrEqualToTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/GreaterThanOrEqualTo");
/** @internal */
const LessThanTypeId = exports.LessThanTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/LessThan");
/** @internal */
const LessThanOrEqualToTypeId = exports.LessThanOrEqualToTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/LessThanOrEqualTo");
/** @internal */
const IntTypeId = exports.IntTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/Int");
/** @internal */
const BetweenTypeId = exports.BetweenTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/Between");
/** @internal */
const GreaterThanBigintTypeId = exports.GreaterThanBigintTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/GreaterThanBigint");
/** @internal */
const GreaterThanOrEqualToBigIntTypeId = exports.GreaterThanOrEqualToBigIntTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/GreaterThanOrEqualToBigint");
/** @internal */
const LessThanBigIntTypeId = exports.LessThanBigIntTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/LessThanBigint");
/** @internal */
const LessThanOrEqualToBigIntTypeId = exports.LessThanOrEqualToBigIntTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/LessThanOrEqualToBigint");
/** @internal */
const BetweenBigintTypeId = exports.BetweenBigintTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/BetweenBigint");
/** @internal */
const MinLengthTypeId = exports.MinLengthTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/MinLength");
/** @internal */
const MaxLengthTypeId = exports.MaxLengthTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/MaxLength");
/** @internal */
const LengthTypeId = exports.LengthTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/Length");
/** @internal */
const MinItemsTypeId = exports.MinItemsTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/MinItems");
/** @internal */
const MaxItemsTypeId = exports.MaxItemsTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/MaxItems");
/** @internal */
const ItemsCountTypeId = exports.ItemsCountTypeId = /*#__PURE__*/Symbol.for("@effect/schema/TypeId/ItemsCount");
