// @effect/schema/dist/esm/internal/errors.js
import * as util_ from "./util.js";
/** @internal */
export const getDuplicatePropertySignatureErrorMessage = name => `Duplicate property signature ${util_.formatUnknown(name)}`;
/** @internal */
export const getDuplicateIndexSignatureErrorMessage = name => `Duplicate index signature for type \`${name}\``;
/** @internal */
export const getIndexSignatureParameterErrorMessage = "An index signature parameter type must be `string`, `symbol`, a template literal type or a refinement of the previous types";
/** @internal */
export const getRequiredElementFollowinAnOptionalElementErrorMessage = "A required element cannot follow an optional element. ts(1257)";
/** @internal */
export const getDuplicatePropertySignatureTransformationErrorMessage = name => `Duplicate property signature transformation ${util_.formatUnknown(name)}`;
/** @internal */
export const getArbitraryErrorMessage = message => `cannot build an Arbitrary for ${message}`;
/** @internal */
export const getPrettyErrorMessage = message => `cannot build a Pretty for ${message}`;
/** @internal */
export const getEquivalenceErrorMessage = message => `cannot build an Equivalence for ${message}`;
/** @internal */
export const getAPIErrorMessage = (api, message) => `${api}: ${message}`;
