// effect/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Runtime = exports.Resource = exports.RequestResolver = exports.RequestBlock = exports.Request = exports.Reloadable = exports.RegExp = exports.Ref = exports.RedBlackTree = exports.Record = exports.Readable = exports.RateLimiter = exports.Random = exports.Queue = exports.PubSub = exports.PrimaryKey = exports.Predicate = exports.Pool = exports.Pipeable = exports.Ordering = exports.Order = exports.Option = exports.Number = exports.NonEmptyIterable = exports.MutableRef = exports.MutableQueue = exports.MutableList = exports.MutableHashSet = exports.MutableHashMap = exports.ModuleVersion = exports.MetricState = exports.MetricRegistry = exports.MetricPolling = exports.MetricPair = exports.MetricLabel = exports.MetricKeyType = exports.MetricKey = exports.MetricHook = exports.MetricBoundaries = exports.Metric = exports.MergeStrategy = exports.MergeState = exports.MergeDecision = exports.Match = exports.ManagedRuntime = exports.Logger = exports.LogSpan = exports.LogLevel = exports.List = exports.Layer = exports.KeyedPool = exports.Iterable = exports.Inspectable = exports.HashSet = exports.HashMap = exports.Hash = exports.HKT = exports.GroupBy = exports.GlobalValue = exports.Function = exports.FiberStatus = exports.FiberSet = exports.FiberRefsPatch = exports.FiberRefs = exports.FiberRef = exports.FiberMap = exports.FiberId = exports.FiberHandle = exports.Fiber = exports.Exit = exports.ExecutionStrategy = exports.Equivalence = exports.Equal = exports.Encoding = exports.Either = exports.Effectable = exports.Effect = exports.Duration = exports.Differ = exports.Deferred = exports.DefaultServices = exports.Data = exports.Cron = exports.Context = exports.Console = exports.ConfigProviderPathPatch = exports.ConfigProvider = exports.ConfigError = exports.Config = exports.Clock = exports.Chunk = exports.ChildExecutorDecision = exports.Channel = exports.Cause = exports.Cache = exports.Brand = exports.Boolean = exports.BigInt = exports.BigDecimal = exports.Array = void 0;
exports.Utils = exports.UpstreamPullStrategy = exports.UpstreamPullRequest = exports.Unify = exports.Types = exports.Tuple = exports.Trie = exports.Tracer = exports.TestSized = exports.TestServices = exports.TestLive = exports.TestContext = exports.TestConfig = exports.TestClock = exports.TestAnnotations = exports.TestAnnotationMap = exports.TestAnnotation = exports.Take = exports.TSet = exports.TSemaphore = exports.TRef = exports.TReentrantLock = exports.TRandom = exports.TQueue = exports.TPubSub = exports.TPriorityQueue = exports.TMap = exports.TDeferred = exports.TArray = exports.SynchronizedRef = exports.Symbol = exports.Supervisor = exports.SubscriptionRef = exports.Subscribable = exports.Struct = exports.String = exports.Streamable = exports.StreamHaltStrategy = exports.StreamEmit = exports.Stream = exports.SortedSet = exports.SortedMap = exports.Sink = exports.SingleProducerAsyncInput = exports.Secret = exports.ScopedRef = exports.ScopedCache = exports.Scope = exports.Scheduler = exports.ScheduleIntervals = exports.ScheduleInterval = exports.ScheduleDecision = exports.Schedule = exports.STM = exports.RuntimeFlagsPatch = exports.RuntimeFlags = void 0;
Object.defineProperty(exports, "absurd", {
  enumerable: true,
  get: function () {
    return _Function.absurd;
  }
});
Object.defineProperty(exports, "flow", {
  enumerable: true,
  get: function () {
    return _Function.flow;
  }
});
Object.defineProperty(exports, "hole", {
  enumerable: true,
  get: function () {
    return _Function.hole;
  }
});
Object.defineProperty(exports, "identity", {
  enumerable: true,
  get: function () {
    return _Function.identity;
  }
});
Object.defineProperty(exports, "pipe", {
  enumerable: true,
  get: function () {
    return _Function.pipe;
  }
});
Object.defineProperty(exports, "unsafeCoerce", {
  enumerable: true,
  get: function () {
    return _Function.unsafeCoerce;
  }
});
var _Function = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Function.js"));
exports.Function = _Function;
var _Array = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Array.js"));
exports.Array = _Array;
var _BigDecimal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./BigDecimal.js"));
exports.BigDecimal = _BigDecimal;
var _BigInt = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./BigInt.js"));
exports.BigInt = _BigInt;
var _Boolean = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Boolean.js"));
exports.Boolean = _Boolean;
var _Brand = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Brand.js"));
exports.Brand = _Brand;
var _Cache = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Cache.js"));
exports.Cache = _Cache;
var _Cause = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Cause.js"));
exports.Cause = _Cause;
var _Channel = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Channel.js"));
exports.Channel = _Channel;
var _ChildExecutorDecision = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ChildExecutorDecision.js"));
exports.ChildExecutorDecision = _ChildExecutorDecision;
var _Chunk = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Chunk.js"));
exports.Chunk = _Chunk;
var _Clock = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Clock.js"));
exports.Clock = _Clock;
var _Config = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Config.js"));
exports.Config = _Config;
var _ConfigError = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ConfigError.js"));
exports.ConfigError = _ConfigError;
var _ConfigProvider = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ConfigProvider.js"));
exports.ConfigProvider = _ConfigProvider;
var _ConfigProviderPathPatch = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ConfigProviderPathPatch.js"));
exports.ConfigProviderPathPatch = _ConfigProviderPathPatch;
var _Console = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Console.js"));
exports.Console = _Console;
var _Context = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Context.js"));
exports.Context = _Context;
var _Cron = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Cron.js"));
exports.Cron = _Cron;
var _Data = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Data.js"));
exports.Data = _Data;
var _DefaultServices = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./DefaultServices.js"));
exports.DefaultServices = _DefaultServices;
var _Deferred = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Deferred.js"));
exports.Deferred = _Deferred;
var _Differ = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Differ.js"));
exports.Differ = _Differ;
var _Duration = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Duration.js"));
exports.Duration = _Duration;
var _Effect = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Effect.js"));
exports.Effect = _Effect;
var _Effectable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Effectable.js"));
exports.Effectable = _Effectable;
var _Either = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Either.js"));
exports.Either = _Either;
var _Encoding = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Encoding.js"));
exports.Encoding = _Encoding;
var _Equal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Equal.js"));
exports.Equal = _Equal;
var _Equivalence = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Equivalence.js"));
exports.Equivalence = _Equivalence;
var _ExecutionStrategy = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ExecutionStrategy.js"));
exports.ExecutionStrategy = _ExecutionStrategy;
var _Exit = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Exit.js"));
exports.Exit = _Exit;
var _Fiber = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Fiber.js"));
exports.Fiber = _Fiber;
var _FiberHandle = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./FiberHandle.js"));
exports.FiberHandle = _FiberHandle;
var _FiberId = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./FiberId.js"));
exports.FiberId = _FiberId;
var _FiberMap = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./FiberMap.js"));
exports.FiberMap = _FiberMap;
var _FiberRef = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./FiberRef.js"));
exports.FiberRef = _FiberRef;
var _FiberRefs = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./FiberRefs.js"));
exports.FiberRefs = _FiberRefs;
var _FiberRefsPatch = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./FiberRefsPatch.js"));
exports.FiberRefsPatch = _FiberRefsPatch;
var _FiberSet = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./FiberSet.js"));
exports.FiberSet = _FiberSet;
var _FiberStatus = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./FiberStatus.js"));
exports.FiberStatus = _FiberStatus;
var _GlobalValue = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./GlobalValue.js"));
exports.GlobalValue = _GlobalValue;
var _GroupBy = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./GroupBy.js"));
exports.GroupBy = _GroupBy;
var _HKT = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./HKT.js"));
exports.HKT = _HKT;
var _Hash = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Hash.js"));
exports.Hash = _Hash;
var _HashMap = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./HashMap.js"));
exports.HashMap = _HashMap;
var _HashSet = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./HashSet.js"));
exports.HashSet = _HashSet;
var _Inspectable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Inspectable.js"));
exports.Inspectable = _Inspectable;
var _Iterable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Iterable.js"));
exports.Iterable = _Iterable;
var _KeyedPool = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./KeyedPool.js"));
exports.KeyedPool = _KeyedPool;
var _Layer = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Layer.js"));
exports.Layer = _Layer;
var _List = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./List.js"));
exports.List = _List;
var _LogLevel = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./LogLevel.js"));
exports.LogLevel = _LogLevel;
var _LogSpan = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./LogSpan.js"));
exports.LogSpan = _LogSpan;
var _Logger = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Logger.js"));
exports.Logger = _Logger;
var _ManagedRuntime = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ManagedRuntime.js"));
exports.ManagedRuntime = _ManagedRuntime;
var _Match = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Match.js"));
exports.Match = _Match;
var _MergeDecision = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MergeDecision.js"));
exports.MergeDecision = _MergeDecision;
var _MergeState = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MergeState.js"));
exports.MergeState = _MergeState;
var _MergeStrategy = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MergeStrategy.js"));
exports.MergeStrategy = _MergeStrategy;
var _Metric = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Metric.js"));
exports.Metric = _Metric;
var _MetricBoundaries = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MetricBoundaries.js"));
exports.MetricBoundaries = _MetricBoundaries;
var _MetricHook = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MetricHook.js"));
exports.MetricHook = _MetricHook;
var _MetricKey = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MetricKey.js"));
exports.MetricKey = _MetricKey;
var _MetricKeyType = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MetricKeyType.js"));
exports.MetricKeyType = _MetricKeyType;
var _MetricLabel = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MetricLabel.js"));
exports.MetricLabel = _MetricLabel;
var _MetricPair = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MetricPair.js"));
exports.MetricPair = _MetricPair;
var _MetricPolling = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MetricPolling.js"));
exports.MetricPolling = _MetricPolling;
var _MetricRegistry = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MetricRegistry.js"));
exports.MetricRegistry = _MetricRegistry;
var _MetricState = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MetricState.js"));
exports.MetricState = _MetricState;
var _ModuleVersion = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ModuleVersion.js"));
exports.ModuleVersion = _ModuleVersion;
var _MutableHashMap = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MutableHashMap.js"));
exports.MutableHashMap = _MutableHashMap;
var _MutableHashSet = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MutableHashSet.js"));
exports.MutableHashSet = _MutableHashSet;
var _MutableList = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MutableList.js"));
exports.MutableList = _MutableList;
var _MutableQueue = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MutableQueue.js"));
exports.MutableQueue = _MutableQueue;
var _MutableRef = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./MutableRef.js"));
exports.MutableRef = _MutableRef;
var _NonEmptyIterable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./NonEmptyIterable.js"));
exports.NonEmptyIterable = _NonEmptyIterable;
var _Number = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Number.js"));
exports.Number = _Number;
var _Option = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Option.js"));
exports.Option = _Option;
var _Order = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Order.js"));
exports.Order = _Order;
var _Ordering = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Ordering.js"));
exports.Ordering = _Ordering;
var _Pipeable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Pipeable.js"));
exports.Pipeable = _Pipeable;
var _Pool = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Pool.js"));
exports.Pool = _Pool;
var _Predicate = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Predicate.js"));
exports.Predicate = _Predicate;
var _PrimaryKey = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./PrimaryKey.js"));
exports.PrimaryKey = _PrimaryKey;
var _PubSub = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./PubSub.js"));
exports.PubSub = _PubSub;
var _Queue = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Queue.js"));
exports.Queue = _Queue;
var _Random = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Random.js"));
exports.Random = _Random;
var _RateLimiter = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./RateLimiter.js"));
exports.RateLimiter = _RateLimiter;
var _Readable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Readable.js"));
exports.Readable = _Readable;
var _Record = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Record.js"));
exports.Record = _Record;
var _RedBlackTree = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./RedBlackTree.js"));
exports.RedBlackTree = _RedBlackTree;
var _Ref = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Ref.js"));
exports.Ref = _Ref;
var _RegExp = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./RegExp.js"));
exports.RegExp = _RegExp;
var _Reloadable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Reloadable.js"));
exports.Reloadable = _Reloadable;
var _Request = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Request.js"));
exports.Request = _Request;
var _RequestBlock = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./RequestBlock.js"));
exports.RequestBlock = _RequestBlock;
var _RequestResolver = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./RequestResolver.js"));
exports.RequestResolver = _RequestResolver;
var _Resource = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Resource.js"));
exports.Resource = _Resource;
var _Runtime = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Runtime.js"));
exports.Runtime = _Runtime;
var _RuntimeFlags = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./RuntimeFlags.js"));
exports.RuntimeFlags = _RuntimeFlags;
var _RuntimeFlagsPatch = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./RuntimeFlagsPatch.js"));
exports.RuntimeFlagsPatch = _RuntimeFlagsPatch;
var _STM = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./STM.js"));
exports.STM = _STM;
var _Schedule = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Schedule.js"));
exports.Schedule = _Schedule;
var _ScheduleDecision = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ScheduleDecision.js"));
exports.ScheduleDecision = _ScheduleDecision;
var _ScheduleInterval = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ScheduleInterval.js"));
exports.ScheduleInterval = _ScheduleInterval;
var _ScheduleIntervals = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ScheduleIntervals.js"));
exports.ScheduleIntervals = _ScheduleIntervals;
var _Scheduler = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Scheduler.js"));
exports.Scheduler = _Scheduler;
var _Scope = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Scope.js"));
exports.Scope = _Scope;
var _ScopedCache = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ScopedCache.js"));
exports.ScopedCache = _ScopedCache;
var _ScopedRef = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./ScopedRef.js"));
exports.ScopedRef = _ScopedRef;
var _Secret = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Secret.js"));
exports.Secret = _Secret;
var _SingleProducerAsyncInput = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./SingleProducerAsyncInput.js"));
exports.SingleProducerAsyncInput = _SingleProducerAsyncInput;
var _Sink = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Sink.js"));
exports.Sink = _Sink;
var _SortedMap = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./SortedMap.js"));
exports.SortedMap = _SortedMap;
var _SortedSet = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./SortedSet.js"));
exports.SortedSet = _SortedSet;
var _Stream = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Stream.js"));
exports.Stream = _Stream;
var _StreamEmit = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./StreamEmit.js"));
exports.StreamEmit = _StreamEmit;
var _StreamHaltStrategy = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./StreamHaltStrategy.js"));
exports.StreamHaltStrategy = _StreamHaltStrategy;
var _Streamable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Streamable.js"));
exports.Streamable = _Streamable;
var _String = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./String.js"));
exports.String = _String;
var _Struct = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Struct.js"));
exports.Struct = _Struct;
var _Subscribable = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Subscribable.js"));
exports.Subscribable = _Subscribable;
var _SubscriptionRef = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./SubscriptionRef.js"));
exports.SubscriptionRef = _SubscriptionRef;
var _Supervisor = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Supervisor.js"));
exports.Supervisor = _Supervisor;
var _Symbol = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Symbol.js"));
exports.Symbol = _Symbol;
var _SynchronizedRef = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./SynchronizedRef.js"));
exports.SynchronizedRef = _SynchronizedRef;
var _TArray = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TArray.js"));
exports.TArray = _TArray;
var _TDeferred = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TDeferred.js"));
exports.TDeferred = _TDeferred;
var _TMap = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TMap.js"));
exports.TMap = _TMap;
var _TPriorityQueue = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TPriorityQueue.js"));
exports.TPriorityQueue = _TPriorityQueue;
var _TPubSub = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TPubSub.js"));
exports.TPubSub = _TPubSub;
var _TQueue = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TQueue.js"));
exports.TQueue = _TQueue;
var _TRandom = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TRandom.js"));
exports.TRandom = _TRandom;
var _TReentrantLock = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TReentrantLock.js"));
exports.TReentrantLock = _TReentrantLock;
var _TRef = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TRef.js"));
exports.TRef = _TRef;
var _TSemaphore = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TSemaphore.js"));
exports.TSemaphore = _TSemaphore;
var _TSet = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TSet.js"));
exports.TSet = _TSet;
var _Take = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Take.js"));
exports.Take = _Take;
var _TestAnnotation = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TestAnnotation.js"));
exports.TestAnnotation = _TestAnnotation;
var _TestAnnotationMap = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TestAnnotationMap.js"));
exports.TestAnnotationMap = _TestAnnotationMap;
var _TestAnnotations = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TestAnnotations.js"));
exports.TestAnnotations = _TestAnnotations;
var _TestClock = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TestClock.js"));
exports.TestClock = _TestClock;
var _TestConfig = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TestConfig.js"));
exports.TestConfig = _TestConfig;
var _TestContext = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TestContext.js"));
exports.TestContext = _TestContext;
var _TestLive = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TestLive.js"));
exports.TestLive = _TestLive;
var _TestServices = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TestServices.js"));
exports.TestServices = _TestServices;
var _TestSized = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./TestSized.js"));
exports.TestSized = _TestSized;
var _Tracer = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Tracer.js"));
exports.Tracer = _Tracer;
var _Trie = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Trie.js"));
exports.Trie = _Trie;
var _Tuple = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Tuple.js"));
exports.Tuple = _Tuple;
var _Types = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Types.js"));
exports.Types = _Types;
var _Unify = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Unify.js"));
exports.Unify = _Unify;
var _UpstreamPullRequest = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./UpstreamPullRequest.js"));
exports.UpstreamPullRequest = _UpstreamPullRequest;
var _UpstreamPullStrategy = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./UpstreamPullStrategy.js"));
exports.UpstreamPullStrategy = _UpstreamPullStrategy;
var _Utils = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./Utils.js"));
exports.Utils = _Utils;
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
