// effect/dist/cjs/MergeDecision.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.match = exports.isMergeDecision = exports.MergeDecisionTypeId = exports.Done = exports.AwaitConst = exports.Await = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./internal/channel/mergeDecision.js"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
/**
 * @since 2.0.0
 * @category symbols
 */
const MergeDecisionTypeId = exports.MergeDecisionTypeId = internal.MergeDecisionTypeId;
/**
 * @since 2.0.0
 * @category constructors
 */
const Done = exports.Done = internal.Done;
/**
 * @since 2.0.0
 * @category constructors
 */
const Await = exports.Await = internal.Await;
/**
 * @since 2.0.0
 * @category constructors
 */
const AwaitConst = exports.AwaitConst = internal.AwaitConst;
/**
 * Returns `true` if the specified value is a `MergeDecision`, `false`
 * otherwise.
 *
 * @since 2.0.0
 * @category refinements
 */
const isMergeDecision = exports.isMergeDecision = internal.isMergeDecision;
/**
 * @since 2.0.0
 * @category folding
 */
const match = exports.match = internal.match;
