// effect/dist/cjs/MetricState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.summary = exports.isSummaryState = exports.isMetricState = exports.isHistogramState = exports.isGaugeState = exports.isFrequencyState = exports.isCounterState = exports.histogram = exports.gauge = exports.frequency = exports.counter = exports.SummaryStateTypeId = exports.MetricStateTypeId = exports.HistogramStateTypeId = exports.GaugeStateTypeId = exports.FrequencyStateTypeId = exports.CounterStateTypeId = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./internal/metric/state.js"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
/**
 * @since 2.0.0
 * @category symbols
 */
const MetricStateTypeId = exports.MetricStateTypeId = internal.MetricStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const CounterStateTypeId = exports.CounterStateTypeId = internal.CounterStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const FrequencyStateTypeId = exports.FrequencyStateTypeId = internal.FrequencyStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const GaugeStateTypeId = exports.GaugeStateTypeId = internal.GaugeStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const HistogramStateTypeId = exports.HistogramStateTypeId = internal.HistogramStateTypeId;
/**
 * @since 2.0.0
 * @category symbols
 */
const SummaryStateTypeId = exports.SummaryStateTypeId = internal.SummaryStateTypeId;
/**
 * @since 2.0.0
 * @category constructors
 */
const counter = exports.counter = internal.counter;
/**
 * @since 2.0.0
 * @category constructors
 */
const frequency = exports.frequency = internal.frequency;
/**
 * @since 2.0.0
 * @category constructors
 */
const gauge = exports.gauge = internal.gauge;
/**
 * @since 2.0.0
 * @category constructors
 */
const histogram = exports.histogram = internal.histogram;
/**
 * @since 2.0.0
 * @category constructors
 */
const summary = exports.summary = internal.summary;
/**
 * @since 2.0.0
 * @category refinements
 */
const isMetricState = exports.isMetricState = internal.isMetricState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isCounterState = exports.isCounterState = internal.isCounterState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isFrequencyState = exports.isFrequencyState = internal.isFrequencyState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isGaugeState = exports.isGaugeState = internal.isGaugeState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isHistogramState = exports.isHistogramState = internal.isHistogramState;
/**
 * @since 2.0.0
 * @category refinements
 */
const isSummaryState = exports.isSummaryState = internal.isSummaryState;
