// effect/dist/cjs/Secret.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.value = exports.unsafeWipe = exports.make = exports.isSecret = exports.fromString = exports.fromIterable = exports.SecretTypeId = void 0;
var InternalSecret = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./internal/secret.js"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
/**
 * @since 2.0.0
 * @category symbols
 */
const SecretTypeId = exports.SecretTypeId = InternalSecret.SecretTypeId;
/**
 * @since 2.0.0
 * @category refinements
 */
const isSecret = exports.isSecret = InternalSecret.isSecret;
/**
 * @since 2.0.0
 * @category constructors
 */
const make = exports.make = InternalSecret.make;
/**
 * @since 2.0.0
 * @category constructors
 */
const fromIterable = exports.fromIterable = InternalSecret.fromIterable;
/**
 * @since 2.0.0
 * @category constructors
 */
const fromString = exports.fromString = InternalSecret.fromString;
/**
 * @since 2.0.0
 * @category getters
 */
const value = exports.value = InternalSecret.value;
/**
 * @since 2.0.0
 * @category unsafe
 */
const unsafeWipe = exports.unsafeWipe = InternalSecret.unsafeWipe;
