// effect/dist/cjs/StreamHaltStrategy.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.match = exports.isRight = exports.isLeft = exports.isEither = exports.isBoth = exports.fromInput = exports.Right = exports.Left = exports.Either = exports.Both = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./internal/stream/haltStrategy.js"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
/**
 * @since 2.0.0
 */

/**
 * @since 2.0.0
 * @category constructors
 */
const Left = exports.Left = internal.Left;
/**
 * @since 2.0.0
 * @category constructors
 */
const Right = exports.Right = internal.Right;
/**
 * @since 2.0.0
 * @category constructors
 */
const Both = exports.Both = internal.Both;
/**
 * @since 2.0.0
 * @category constructors
 */
const Either = exports.Either = internal.Either;
/**
 * @since 2.0.0
 * @category constructors
 */
const fromInput = exports.fromInput = internal.fromInput;
/**
 * @since 2.0.0
 * @category refinements
 */
const isLeft = exports.isLeft = internal.isLeft;
/**
 * @since 2.0.0
 * @category refinements
 */
const isRight = exports.isRight = internal.isRight;
/**
 * @since 2.0.0
 * @category refinements
 */
const isBoth = exports.isBoth = internal.isBoth;
/**
 * @since 2.0.0
 * @category refinements
 */
const isEither = exports.isEither = internal.isEither;
/**
 * Folds over the specified `HaltStrategy` using the provided case functions.
 *
 * @since 2.0.0
 * @category folding
 */
const match = exports.match = internal.match;
