// effect/dist/cjs/Tracer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tracerWith = exports.make = exports.externalSpan = exports.TracerTypeId = exports.Tracer = exports.ParentSpan = void 0;
var defaultServices = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./internal/defaultServices.js"));
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./internal/tracer.js"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
/**
 * @since 2.0.0
 */
const TracerTypeId = exports.TracerTypeId = internal.TracerTypeId;
/**
 * @since 2.0.0
 * @category tags
 */
const ParentSpan = exports.ParentSpan = internal.spanTag;
/**
 * @since 2.0.0
 * @category tags
 */
const Tracer = exports.Tracer = internal.tracerTag;
/**
 * @since 2.0.0
 * @category constructors
 */
const make = exports.make = internal.make;
/**
 * @since 2.0.0
 * @category constructors
 */
const externalSpan = exports.externalSpan = internal.externalSpan;
/**
 * @since 2.0.0
 * @category constructors
 */
const tracerWith = exports.tracerWith = defaultServices.tracerWith;
