// effect/dist/cjs/TSemaphore.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withPermitsScoped = exports.withPermits = exports.withPermitScoped = exports.withPermit = exports.unsafeMake = exports.releaseN = exports.release = exports.make = exports.available = exports.acquireN = exports.acquire = exports.TSemaphoreTypeId = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./internal/stm/tSemaphore.js"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
/**
 * @since 2.0.0
 */

/**
 * @since 2.0.0
 * @category symbols
 */
const TSemaphoreTypeId = exports.TSemaphoreTypeId = internal.TSemaphoreTypeId;
/**
 * @since 2.0.0
 * @category mutations
 */
const acquire = exports.acquire = internal.acquire;
/**
 * @since 2.0.0
 * @category mutations
 */
const acquireN = exports.acquireN = internal.acquireN;
/**
 * @since 2.0.0
 * @category getters
 */
const available = exports.available = internal.available;
/**
 * @since 2.0.0
 * @category constructors
 */
const make = exports.make = internal.make;
/**
 * @since 2.0.0
 * @category mutations
 */
const release = exports.release = internal.release;
/**
 * @since 2.0.0
 * @category mutations
 */
const releaseN = exports.releaseN = internal.releaseN;
/**
 * @since 2.0.0
 * @category mutations
 */
const withPermit = exports.withPermit = internal.withPermit;
/**
 * @since 2.0.0
 * @category mutations
 */
const withPermits = exports.withPermits = internal.withPermits;
/**
 * @since 2.0.0
 * @category mutations
 */
const withPermitScoped = exports.withPermitScoped = internal.withPermitScoped;
/**
 * @since 2.0.0
 * @category mutations
 */
const withPermitsScoped = exports.withPermitsScoped = internal.withPermitsScoped;
/**
 * @since 2.0.0
 * @category unsafe
 */
const unsafeMake = exports.unsafeMake = internal.unsafeMakeSemaphore;
