// @ark/schema/out/refinements/minLength.js
import { implementNode } from "../shared/implement.js";
import { $ark } from "../shared/registry.js";
import { BaseRange, parseExclusiveKey } from "./range.js";
const implementation = implementNode({
    kind: "minLength",
    collapsibleKey: "rule",
    hasAssociatedError: true,
    keys: {
        rule: {},
        exclusive: parseExclusiveKey
    },
    normalize: schema => typeof schema === "number" ? { rule: schema } : schema,
    defaults: {
        description: node => node.exclusive ?
            node.rule === 0 ?
                "non-empty"
                : `more than length ${node.rule}`
            : node.rule === 1 ? "non-empty"
                : `at least length ${node.rule}`,
        // avoid default message like "must be non-empty (was 0)"
        actual: data => (data.length === 0 ? null : `${data.length}`)
    },
    intersections: {
        minLength: (l, r) => (l.isStricterThan(r) ? l : r)
    }
});
export class MinLengthNode extends BaseRange {
    impliedBasis = $ark.intrinsic.lengthBoundable.internal;
    traverseAllows = this.exclusive ?
        data => data.length > this.rule
        : data => data.length >= this.rule;
}
export const MinLength = {
    implementation,
    Node: MinLengthNode
};
