// arktype/dist/cjs/main.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    scope: ()=>_scopeJs.scope,
    type: ()=>_arkJs.type,
    ark: ()=>_arkJs.ark,
    arkScope: ()=>_arkJs.arkScope,
    arrayOf: ()=>_expressionsJs.arrayOf,
    instanceOf: ()=>_expressionsJs.instanceOf,
    intersection: ()=>_expressionsJs.intersection,
    keyOf: ()=>_expressionsJs.keyOf,
    morph: ()=>_expressionsJs.morph,
    narrow: ()=>_expressionsJs.narrow,
    union: ()=>_expressionsJs.union,
    valueOf: ()=>_expressionsJs.valueOf,
    jsObjectsScope: ()=>_jsObjectsJs.jsObjectsScope,
    tsKeywordsScope: ()=>_tsKeywordsJs.tsKeywordsScope,
    validationScope: ()=>_validationJs.validationScope,
    Problems: ()=>_problemsJs.Problems,
    Problem: ()=>_problemsJs.Problem,
    parseConfigTuple: ()=>_configJs.parseConfigTuple
});
const _scopeJs = require("./scopes/scope.js");
const _arkJs = require("./scopes/ark.js");
const _expressionsJs = require("./scopes/expressions.js");
const _jsObjectsJs = require("./scopes/jsObjects.js");
const _tsKeywordsJs = require("./scopes/tsKeywords.js");
const _validationJs = require("./scopes/validation/validation.js");
const _problemsJs = require("./traverse/problems.js");
const _configJs = require("./parse/ast/config.js");
