// arktype/dist/cjs/scopes/ark.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    arkScope: ()=>arkScope,
    ark: ()=>ark,
    scopes: ()=>scopes,
    spaces: ()=>spaces,
    type: ()=>type
});
const _jsObjectsJs = require("./jsObjects.js");
const _scopeJs = require("./scope.js");
const _tsKeywordsJs = require("./tsKeywords.js");
const _validationJs = require("./validation/validation.js");
const arkScope = (0, _scopeJs.scope)({}, {
    name: "standard",
    includes: [
        _tsKeywordsJs.tsKeywords,
        _jsObjectsJs.jsObjects,
        _validationJs.validation
    ],
    standard: false
});
const ark = arkScope.compile();
const scopes = {
    root: _scopeJs.rootScope,
    tsKeywords: _tsKeywordsJs.tsKeywordsScope,
    jsObjects: _jsObjectsJs.jsObjectsScope,
    validation: _validationJs.validationScope,
    ark: arkScope
};
const spaces = {
    tsKeywords: _tsKeywordsJs.tsKeywords,
    jsObjects: _jsObjectsJs.jsObjects,
    validation: _validationJs.validation,
    ark
};
const type = arkScope.type;
