// arktype/dist/cjs/scopes/jsObjects.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    jsObjectsScope: ()=>jsObjectsScope,
    jsObjects: ()=>jsObjects
});
const _scopeJs = require("./scope.js");
const jsObjectsScope = (0, _scopeJs.scope)({
    Function: [
        "node",
        {
            object: {
                class: Function
            }
        }
    ],
    Date: [
        "node",
        {
            object: {
                class: Date
            }
        }
    ],
    Error: [
        "node",
        {
            object: {
                class: Error
            }
        }
    ],
    Map: [
        "node",
        {
            object: {
                class: Map
            }
        }
    ],
    RegExp: [
        "node",
        {
            object: {
                class: RegExp
            }
        }
    ],
    Set: [
        "node",
        {
            object: {
                class: Set
            }
        }
    ],
    WeakMap: [
        "node",
        {
            object: {
                class: WeakMap
            }
        }
    ],
    WeakSet: [
        "node",
        {
            object: {
                class: WeakSet
            }
        }
    ],
    Promise: [
        "node",
        {
            object: {
                class: Promise
            }
        }
    ]
}, {
    name: "jsObjects",
    standard: false
});
const jsObjects = jsObjectsScope.compile();
