// arktype/dist/cjs/scopes/tsKeywords.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    tsKeywordsScope: ()=>tsKeywordsScope,
    tsKeywords: ()=>tsKeywords
});
const _scopeJs = require("./scope.js");
const always = {
    bigint: true,
    boolean: true,
    null: true,
    number: true,
    object: true,
    string: true,
    symbol: true,
    undefined: true
};
const tsKeywordsScope = (0, _scopeJs.scope)({
    any: [
        "node",
        always
    ],
    bigint: [
        "node",
        {
            bigint: true
        }
    ],
    boolean: [
        "node",
        {
            boolean: true
        }
    ],
    false: [
        "node",
        {
            boolean: {
                value: false
            }
        }
    ],
    never: [
        "node",
        {}
    ],
    null: [
        "node",
        {
            null: true
        }
    ],
    number: [
        "node",
        {
            number: true
        }
    ],
    object: [
        "node",
        {
            object: true
        }
    ],
    string: [
        "node",
        {
            string: true
        }
    ],
    symbol: [
        "node",
        {
            symbol: true
        }
    ],
    true: [
        "node",
        {
            boolean: {
                value: true
            }
        }
    ],
    unknown: [
        "node",
        always
    ],
    void: [
        "node",
        {
            undefined: true
        }
    ],
    undefined: [
        "node",
        {
            undefined: true
        }
    ]
}, {
    name: "ts",
    standard: false
});
const tsKeywords = tsKeywordsScope.compile();
